### DISCUSSION TITLE
partial component

### DISCUSSION THREAD

**Discussion by amitu** [fastn-stack/fastn#897] [Ideas & RFCs]
```ftd
-- ftd.text foo: hell
```

creates a partial component called foo, which can be used like

```ftd
-- foo:
color: red
```

Where the rest of the arguments are passed. The properties passed to foo will be merged with properties specified during alias create, and will be passed to ftd.text.

**Comment by amitu** [comment]
```ftd
foo : x -> y -> Component

foo-x: y -> Component
foo-x = f 10


-- foo foo-x:
x: 10

-- foo-x:
y: 20

;; should be allowed because `foo-x` may have specified many properties 
;; and we may want to override a subset of them
-- foo-x:
x: 5
y: 20

-- component foo:
integer x:
integer y:

-- end: foo
```

**Comment by siddhantk232** [comment]
This definitely has the coolness factor to it. It was hard for me to read tho. This can also be extended to records and functions.

**Reply by amitu** [reply]
> It was hard for me to read tho

You want me to explain anything here?

**Reply by siddhantk232** [reply]
No No. I can understand the snippet, it's just harder to read.

```
-- foo foo-x:
x: 10
```

Can also be interpreted as: "`foo-x` is a variable of type `foo` where `foo` is a record that takes an integer header `x`"

**Reply by siddhantk232** [reply]
But that cannot happen as `fastn` will know that `foo` is a component definition and not a record definition.

**Reply by amitu** [reply]
Yeah. This is one weakness / feature, the parsing itself happens after finding the type. This is usually not done in other languages, where type of something is almost always known (e.g. in Rust, say  `8` is "auto int", and exact kind of int or float is decided later on after type checking, we do like that for most things, eg in Rust `'a'` is always a `char`, no matter what's the context around it, unless it is part of a quoted string).
